src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+        private static byte[] s_nicAddress;             // cache the NIC address from the registry
+        private readonly GlobalTransactionsFeature _globalTransactionsFeature;
+        private readonly DataClassificationFeature _dataClassificationFeature;
+        private readonly Utf8Feature _utf8SupportFeature;
+        private readonly SqlDnsCachingFeature _sqlDnsCachingFeature;
+        private readonly SessionRecoveryFeature _sessionRecoveryFeature;
+        private readonly FedAuthFeature _fedAuthFeature;
+        private readonly TceFeature _tceFeature;
+
+        /// <summary>
+        /// Parameterless constructor for the login handler.
+        /// </summary>
+        public LoginHandler()
+        {
+            // Setup the feature extensions.
+            _sessionRecoveryFeature = new SessionRecoveryFeature();
+            _fedAuthFeature = new FedAuthFeature();
+            _tceFeature = new TceFeature();
+            _globalTransactionsFeature = new GlobalTransactionsFeature();
+            _dataClassificationFeature = new DataClassificationFeature();
+            _utf8SupportFeature = new Utf8Feature();
+            _sqlDnsCachingFeature = new SqlDnsCachingFeature();
+        }
+
+        public override async ValueTask Handle(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            LoginHandlerContext loginHandlerContext = new LoginHandlerContext(context);
+            await SendLogin(loginHandlerContext, isAsync, ct).ConfigureAwait(false);
+
+            // TODO: Complete the login by reading data. 
+            // This requires parsing by reading token stream from TDS.
+        }
+
+        private async ValueTask SendLogin(LoginHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+
+            PasswordChangeRequest passwordChangeRequest = context.PasswordChangeRequest;
+
+            TdsEnums.FeatureExtension requestedFeatures = TdsEnums.FeatureExtension.None;
+            FeatureExtensions features = context.Features;
+            if (context.ConnectionOptions.ConnectRetryCount > 0)
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.SessionRecovery;
+                features.SessionRecoveryRequested = true;
+            }
+
+            
+            if (ShouldRequestFedAuth(context))
+            {
+                requestedFeatures |= TdsEnums.FeatureExtension.FedAuth;
+                features.FederatedAuthenticationInfoRequested = true;
+                features.FedAuthFeatureExtensionData =
+                    new FederatedAuthenticationFeatureExtensionData
+                    {
+                        libraryType = TdsEnums.FedAuthLibrary.MSAL,
+                        authentication = context.ConnectionOptions.Authentication,
+                        fedAuthRequiredPreLoginResponse = context.FedAuthNegotiatedInPrelogin
+                    };
+            }
+
+            if (context.AccessTokenInBytes != null)
```

Comment:
Style point here: I'd integrate `context.AccessTokenInBytes` into the return value of `ShouldRequestFedAuth`, then just set `FedAuthFeatureExtensionData` differently based on `context.AccessTokenInBytes`. I'd also consider having two different constructors for `FederatedAuthenticationFeatureExtensionData`.